tests/Temporalio.Tests/Worker/WorkflowWorkerTests.cs

```diff
@@ -4789,6 +4790,215 @@ public async Task ExecuteWorkflowAsync_CoroutinesAfterComplete_GetProcessed()
     }
 
+    public class CallbackActivities
+    {
+        private readonly Func<string, Task<string>> func;
+
+        public CallbackActivities(Func<string, Task<string>> func) => this.func = func;
+
+        [Activity]
+        public Task<string> DoActivityAsync(string param) => func(param);
+    }
+
+    [Workflow]
+    public class SemaphoreWorkflow : IDisposable
+    {
+        private readonly SemaphoreSlim semaSlim = new(1, 1);
+        private readonly Semaphore semaHeavy = new(1, 1);
+
+        [WorkflowRun]
+        public Task RunAsync() => Workflow.WaitConditionAsync(() => CompletedUpdates.Count == 3);
+
+        [WorkflowQuery]
+        public List<string> CompletedUpdates { get; } = new();
+
+        [WorkflowUpdate]
+        public async Task UpdateWithAsyncSlim()
+        {
+            await semaSlim.WaitAsync();
+            try
+            {
+                CompletedUpdates.Add(await Workflow.ExecuteActivityAsync(
+                    (CallbackActivities act) => act.DoActivityAsync(Workflow.CurrentUpdateInfo!.Id),
+                    new() { StartToCloseTimeout = TimeSpan.FromMinutes(10) }));
+            }
+            finally
+            {
+                semaSlim.Release();
+            }
+        }
+
+        [WorkflowUpdate]
+        public async Task UpdateWithSyncSlim()
+        {
+#pragma warning disable CA1849, VSTHRD103 // Intentionally block thread
+            semaSlim.Wait();
+#pragma warning restore CA1849, VSTHRD103
+            try
+            {
+                CompletedUpdates.Add(await Workflow.ExecuteActivityAsync(
+                    (CallbackActivities act) => act.DoActivityAsync(Workflow.CurrentUpdateInfo!.Id),
+                    new() { StartToCloseTimeout = TimeSpan.FromMinutes(10) }));
+            }
+            finally
+            {
+                semaSlim.Release();
+            }
+        }
+
+        [WorkflowUpdate]
+        public async Task UpdateWithUsing()
+        {
+            using var ignore = new Locker(semaSlim);
+            CompletedUpdates.Add(await Workflow.ExecuteActivityAsync(
+                (CallbackActivities act) => act.DoActivityAsync(Workflow.CurrentUpdateInfo!.Id),
+                new() { StartToCloseTimeout = TimeSpan.FromMinutes(10) }));
+        }
+
+        [WorkflowUpdate]
+        public async Task UpdateWithHeavy()
+        {
+            semaHeavy.WaitOne();
+            try
+            {
+                CompletedUpdates.Add(await Workflow.ExecuteActivityAsync(
+                    (CallbackActivities act) => act.DoActivityAsync(Workflow.CurrentUpdateInfo!.Id),
+                    new() { StartToCloseTimeout = TimeSpan.FromMinutes(10) }));
+            }
+            finally
+            {
+                semaHeavy.Release();
+            }
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
```

Comment:
Ah, I can see why we might want to wrap for convenience.

Reply:
Right, this is a bit annoying (not to mention the other bad calls on `SemaphoreSlim`).